File: src/yoinkc/packaging.py
"""Create tarball from rendered output directory."""

import re
import socket
import tarfile
from datetime import datetime
from pathlib import Path
from typing import Optional


def sanitize_hostname(hostname: str) -> str:
    """Remove characters unsafe for filenames; fall back to 'unknown'."""
    cleaned = re.sub(r"[^\w.-]", "", hostname)
    return cleaned or "unknown"


def _resolve_hostname(host_root: Optional[Path] = None) -> str:
    """Hostname with fallback chain: {host_root}/etc/hostname -> socket -> 'unknown'.

    When running inside a container, socket.gethostname() returns the container
    ID rather than the inspected host's name. Reading from {host_root}/etc/hostname
    (where the host filesystem is bind-mounted) gives the correct hostname.
    """
    if host_root is not None:
        try:
            name = (host_root / "etc" / "hostname").read_text().splitlines()[0].strip()
            if name:
                return name
        except (OSError, IndexError):
            pass
    try:
        name = socket.gethostname()
        if name:
            return name
    except OSError:
        pass
    return "unknown"


def get_output_stamp(
    host_root: Optional[Path] = None,
    hostname: Optional[str] = None,
) -> str:
    """Return 'HOSTNAME-YYYYMMDD-HHMMSS' stamp for tarball naming.

    If hostname is provided it is used directly (after sanitizing), bypassing
    filesystem and socket resolution. This lets callers supply the hostname
    already captured in the inspection snapshot, which is more reliable than
    re-reading /etc/hostname (empty on RHEL hosts that use hostnamectl).
    """
    if hostname:
        resolved = sanitize_hostname(hostname)
    else:
        resolved = sanitize_hostname(_resolve_hostname(host_root))
    now = datetime.now().strftime("%Y%m%d-%H%M%S")
    return f"{resolved}-{now}"


def create_tarball(source_dir: Path, tarball_path: Path, prefix: str) -> None:
    """Create a gzipped tarball from source_dir with all entries under prefix/.

    Raises OSError if the tarball cannot be written.
    """
    with tarfile.open(tarball_path, "w:gz") as tf:
        for item in sorted(source_dir.rglob("*")):
            arcname = f"{prefix}/{item.relative_to(source_dir)}"
            tf.add(item, arcname=arcname, recursive=False)

File: cmd/inspectah/internal/pipeline/packaging.go
package pipeline

import (
	"archive/tar"
	"compress/gzip"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"time"
)

var unsafeFilenameRe = regexp.MustCompile(`[^\w.-]`)

// SanitizeHostname removes characters unsafe for filenames.
func SanitizeHostname(hostname string) string {
	cleaned := unsafeFilenameRe.ReplaceAllString(hostname, "")
	if cleaned == "" {
		return "unknown"
	}
	return cleaned
}

// GetOutputStamp returns "HOSTNAME-YYYYMMDD-HHMMSS" for tarball naming.
func GetOutputStamp(hostname string) string {
	resolved := SanitizeHostname(hostname)
	now := time.Now().Format("20060102-150405")
	return fmt.Sprintf("%s-%s", resolved, now)
}

// CreateTarball creates a gzipped tarball from sourceDir with entries
// under prefix/.
func CreateTarball(sourceDir, tarballPath, prefix string) error {
	f, err := os.Create(tarballPath)
	if err != nil {
		return fmt.Errorf("create tarball: %w", err)
	}
	defer f.Close()

	gw := gzip.NewWriter(f)
	defer gw.Close()

	tw := tar.NewWriter(gw)
	defer tw.Close()

	// Collect and sort paths for deterministic output
	var paths []string
	filepath.Walk(sourceDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		paths = append(paths, path)
		return nil
	})
	sort.Strings(paths)

	for _, path := range paths {
		info, err := os.Stat(path)
		if err != nil {
			continue
		}

		rel, err := filepath.Rel(sourceDir, path)
		if err != nil {
			continue
		}
		arcname := filepath.Join(prefix, rel)

		header, err := tar.FileInfoHeader(info, "")
		if err != nil {
			continue
		}
		header.Name = arcname

		if err := tw.WriteHeader(header); err != nil {
			return fmt.Errorf("write header %s: %w", arcname, err)
		}

		if !info.IsDir() {
			f, err := os.Open(path)
			if err != nil {
				continue
			}
			io.Copy(tw, f)
			f.Close()
		}
	}

	return nil
}
